### DISCUSSION TITLE
JIT's watch mode with Create React App v4.x

### DISCUSSION THREAD

**Discussion by bennettdams** [tailwindlabs/tailwindcss#4052] [Help] [answered]
I want to use the JIT mode with the newest Create React App version (v4.0.2) - has anyone already done this?
I got it running with the PostCSS compatibility version, it creates one-off builds like it shoud.
But there's one thing I don't understand: How to start JIT's watch mode together with `react-scripts start`?

Here's my setup:

```
{
  ...
  devDependencies: {
    ...
    "@tailwindcss/postcss7-compat": "^2.1.0",
    "postcss": "^7.0.35",
    "postcss-cli": "^7.1.2",
    "tailwindcss": "npm:@tailwindcss/postcss7-compat@^2.1.0"
    ...
  }
  ...
  scripts: {
    "start": "npm run build:css && react-scripts start",
    "build:css": "TAILWIND_MODE=watch postcss src/styles/tailwind.css -o src/styles/main.css",
    // or
    "build:css": "TAILWIND_MODE=build postcss src/styles/tailwind.css -o src/styles/main.css",
  }
}
```
The `start` script is important here, I have two options:

- if I use `TAILWIND_MODE=watch` in `build:css` at the `start` script, JIT will start in watch mode, but the server itself will never start. I suppose the watcher "overrides" starting the server?
- if I use `TAILWIND_MODE=build` (instead of `watch`) in `build:css`, the watcher does not start, but the server does. But then new classes used in the code won't be added to the CSS file, as the watcher is not running

What am I missing here?

BTW: When using JIT with Next.js (v10.x), there's no need to define anything in the `scripts` of the `package.json` - JIT (also in `watch` mode) just works out of the box. How?

---

Full config:

```js
// tailwind.config.js
module.exports = {
  mode: "jit",
  purge: ["./src/**/*.{js,jsx,ts,tsx}", "./public/index.html"],
  darkMode: false, // or 'media' or 'class'
  theme: {
    extend: {},
  },
  variants: {
    extend: {},
  },
  plugins: [require("@tailwindcss/forms")],
};

```

```js
// postcss.config.js
module.exports = {
  plugins: {
    tailwindcss: {},
    autoprefixer: {},
  },
};


```

**Comment by dance2die** [comment, marked as answer]
> How to start JIT's watch mode together with react-scripts start?

As I was creating a [CRA template](https://github.com/dance2die/cra-template-tailwindcss-typescript), I got around the issue by running `react-start`
and TW's watch process in parallel (`run-p`) using [npm-run-all](https://www.npmjs.com/package/npm-run-all).
(I initially used [CRACO](https://github.com/gsoft-inc/craco) but ended up with this solution)
 
CRA template.json: https://github.com/dance2die/cra-template-tailwindcss-typescript/blob/main/template.json

```json
{
  "package": {
        ...
    },
    "scripts": {
      "build:css": "cross-env TAILWIND_MODE=build NODE_ENV=production postcss src/styles/tailwind.css -o src/styles/index.css",
      "watch:css": "cross-env TAILWIND_MODE=watch NODE_ENV=development postcss src/styles/tailwind.css -o src/styles/index.css --watch",
      "react-scripts:start": "sleep 5 && react-scripts start",
      "react-scripts:build": "react-scripts build",
      "start": "run-p watch:css react-scripts:start",
      "build": "run-s build:css react-scripts:build",
      "test": "react-scripts test",
      "eject": "react-scripts eject"
    }
  }
}
```

You might wonder about `sleep 5` for `"react-scripts:start": "sleep 5 && react-scripts start",`.
It's a magic wait time for `react-scripts` to wait until TW finishes the initial build (might not needed for JIT mode).

You can check out the CRA template [source code](https://github.com/dance2die/cra-template-tailwindcss-typescript) 
or try out the template, [cra-template-tailwindcss-typescript](https://www.npmjs.com/package/cra-template-tailwindcss-typescript).
(You can enable the JIT mode by uncommenting `mode: 'jit'` in `tailwind.config.js`)

![cra jit](https://user-images.githubusercontent.com/8465237/114421160-01506880-9b83-11eb-894f-c0110f9a7b2d.gif)

**Reply by bennettdams** [reply]
Thanks for this workaround. I honestly don't like the amount of manual "configuration" for getting this to work.
What would CRA need to provide so this would be easier to configure? Maybe we could request this change or it is upcoming anyways.
Again, with Next.js the JIT compiler works out of the box.

**Reply by adamwathan** [reply]
I would recommend moving away from CRA and using Vite instead. CRA doesn't let you customize your PostCSS configuration and also hasn't updated to PostCSS 8 yet which makes everything a lot more complicated. Vite is a much nicer experience, you will be blown away 👍🏻

**Reply by naivefun** [reply]
@adamwathan yet Vite does not support webpack5 module federation, which might be support in the next CRA version.

**Reply by Saoirse-Derry** [reply]
Can JIT be switched off in tailwindcss? We've upgraded recently and now have multiple storybook, module-federation projects that nolonger work (Tailwindcss 3, webpack5, nextjs 12)

**Comment by bagsa** [comment]
- [x] [v]()

**Reply by bagsa** [reply]
- [ ] The

**Reply by bagsa** [reply]
****

**Comment by lirbank** [comment]
Not sure if this is related but I found a solution to getting Tailwind JIT work with Storybook hot reload (dev mode), posted here: https://github.com/tailwindlabs/tailwindcss/discussions/6570#discussioncomment-2569691
